### ISSUE TITLE
CORS error in KVS Browser-based Ingestion

### ISSUE THREAD

**Issue by MuhammedBeyhanUS** [aws-samples/amazon-kinesis-video-streams-demos#143]
Hi,
i am trying to implement KVS Browser-based Ingestion and getting CORS error. I'm using mongoose for serving. So far i tried;

-using different browsers.
-using different CORS enabling extensions.
-adding 
```
"Access-Control-Allow-Headers" : "Content-Type", 

"Access-Control-Allow-Origin": "*",

 "Access-Control-Allow-Methods": "OPTIONS,POST,GET,PUT,PATCH"  to the header of putMedia. 
```

The consolo.log is as following;

**Access to fetch at 'https://s-*******.kinesisvideo.eu-west-1.amazonaws.com/putMedia' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.**

**main.js:301 POST https://s-*******.kinesisvideo.eu-west-1.amazonaws.com/putMedia net::ERR_FAILED**

and network output;

**putMedia | CORS error | fetch | main.js:301 | 0 B | 505 ms**

**putMedia | 403 |   | Other | 0 B | 503 ms**

In here [https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_dataplane_PutMedia.html](url) it says 

> “This operation is only available for the AWS SDK for Java. It is not supported in AWS SDKs for other languages.”

would it be the problem?

it is weird that when i use create new stream options, it creates the stream and getDataEndpoint without problem but can't putMedai to endpoint created by himself.

Is there any solution for this CORS Problem? Or any idea?

**Comment by sahilmcompro** [comment]
Hi @MuhammedBeyhanUS, @MushMal,

Have there been any updates on this? The issue is still replicable for me in the attached Browser-JS demo/sample.

@MuhammedBeyhanUS, one more query for you, were you able to use Kinesis Video Stream with some workaround? I am trying with NodeJS Express Server as the intermediatory to do the KVS PutMedia API call. Here, I am getting the "INVALID_MKV_DATA" error.

Thanks, in advance!

**Comment by MuhammedBeyhanUS** [comment]
Hi @sahilmcompro ,

**Comment by MuhammedBeyhanUS** [comment]
hi @sahilmcompro ,
Unfortunately, i could not manage to video stream over KVS as WebRTC, i notice that KVS is not suitable for my case, and i give up on it. The workaround case is something you say, web browser sending the video stream to a dedicated media server and that media server will be using the KVS PutMedia. Since it is too complicated, tiring and costly , now i am using OpenVidu. it would be  good if you have a look at it.
About error, as much as i know, KVS works only MKV video format. The format of video you sending to KVS might not be MKV. 
You are welcome. Have a healthy life.

**Comment by sahilmcompro** [comment]
@MuhammedBeyhanUS, thanks for the prompt reply and I am hoping you are good too.

Regarding third-party solutions, thanks for the suggestion of OpenVidu, I still wanna try with KVS as my use case is exact what KVS provides.

That said, can we keep this issue open? Till someone from the AWS team replies. Thanks!

**Comment by darek-gc** [comment]
Hello,

I also facing CORS problem, browser console:
`[info] load ffmpeg-core
main.js:10 [info] fetch ffmpeg-core.worker.js script
main.js:10 [info] worker object URL=blob:http://localhost:8080/f1c37f39-4e94-4c30-aee5-ef42297e3566
main.js:10 [info] download ffmpeg-core script (~25 MB) from http://localhost:8080/dist/ffmpeg-core.js
main.js:10 [info] initialize ffmpeg-core
main.js:10 [info] ffmpeg-core initialized
main.js:10 [info] ffmpeg-core loaded
main.js:29 RecordRTC version:  5.6.1
main.js:29 started recording video/x-matroska;codecs=avc1 stream.
main.js:57 Using recorderType: D
main.js:124 Passing following config over MediaRecorder API. {type: 'video/x-matroska;codecs=avc1', mimeType: 'video/webm', checkForInactiveTracks: false, initCallback: ƒ}
main.js:29 Recorder state changed: recording
main.js:29 Initialized recorderType: D for output-type: video/x-matroska;codecs=avc1
main.js:29 Stopped recording video/x-matroska;codecs=avc1 stream.
main.js:29 Recorder state changed: stopped
main.js:29 RecordRTC version:  5.6.1
main.js:29 started recording video/x-matroska;codecs=avc1 stream.
main.js:124 Passing following config over MediaRecorder API. {type: 'video/x-matroska;codecs=avc1', mimeType: 'video/webm', initCallback: null, checkForInactiveTracks: false}
main.js:29 Recorder state changed: recording
main.js:29 video/webm -> 0 Bytes
favicon.ico:1 GET http://localhost:8080/favicon.ico 404 (Not Found)
main.js:29 Stopped recording video/x-matroska;codecs=avc1 stream.
main.js:29 Recorder state changed: stopped
main.js:29 RecordRTC version:  5.6.1
main.js:29 started recording video/x-matroska;codecs=avc1 stream.
main.js:124 Passing following config over MediaRecorder API. {type: 'video/x-matroska;codecs=avc1', mimeType: 'video/webm', initCallback: null, checkForInactiveTracks: false}
main.js:29 Recorder state changed: recording
main.js:29 video/webm -> 171 KB
main.js:29 Stopped recording video/x-matroska;codecs=avc1 stream.
main.js:29 Recorder state changed: stopped
main.js:29 RecordRTC version:  5.6.1
main.js:29 started recording video/x-matroska;codecs=avc1 stream.
main.js:124 Passing following config over MediaRecorder API. {type: 'video/x-matroska;codecs=avc1', mimeType: 'video/webm', initCallback: null, checkForInactiveTracks: false}
main.js:29 Recorder state changed: recording
main.js:10 [info] FS.writeFile record
main.js:10 [info] ffmpeg command: ./ffmpeg -nostdin -hide_banner -i record -c:v copy -acodec aac -threads 8 ExampleStream-1638915290108.mkv
main.js:29 video/webm -> 175 KB
main.js:10 [ffmpeg-stderr] Input #0, matroska,webm, from 'record':
main.js:10 [ffmpeg-stderr]   Metadata:
main.js:10 [ffmpeg-stderr]     encoder         : Chrome
main.js:10 [ffmpeg-stderr]   Duration: N/A, start: 0.000000, bitrate: N/A
main.js:10 [ffmpeg-stderr]     Stream #0:0(eng): Audio: opus, 48000 Hz, mono, fltp (default)
main.js:10 [ffmpeg-stderr]     Stream #0:1(eng): Video: h264 (Constrained Baseline), yuv420p(progressive), 640x480 [SAR 1:1 DAR 4:3], 30.30 fps, 30 tbr, 1k tbn, 60 tbc (default)
main.js:10 [ffmpeg-stderr] Stream mapping:
main.js:10 [ffmpeg-stderr]   Stream #0:1 -> #0:0 (copy)
main.js:10 [ffmpeg-stderr]   Stream #0:0 -> #0:1 (opus (native) -> aac (native))
main.js:10 [ffmpeg-stderr] Output #0, matroska, to 'ExampleStream-1638915290108.mkv':
main.js:10 [ffmpeg-stderr]   Metadata:
main.js:10 [ffmpeg-stderr]     encoder         : Lavf58.20.100
main.js:10 [ffmpeg-stderr]     Stream #0:0(eng): Video: h264 (Constrained Baseline) (H264 / 0x34363248), yuv420p(progressive), 640x480 [SAR 1:1 DAR 4:3], q=2-31, 30.30 fps, 30 tbr, 1k tbn, 1k tbc (default)
main.js:10 [ffmpeg-stderr]     Stream #0:1(eng): Audio: aac (LC) ([255][0][0][0] / 0x00FF), 48000 Hz, mono, fltp, 69 kb/s (default)
main.js:10 [ffmpeg-stderr]     Metadata:
main.js:10 [ffmpeg-stderr]       encoder         : Lavc58.35.100 aac
main.js:10 [ffmpeg-stderr] frame=   58 fps=0.0 q=-1.0 Lsize=     176kB time=00:00:01.91 bitrate= 749.8kbits/s speed=33.2x    
main.js:10 [ffmpeg-stderr] video:157kB audio:16kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 1.169377%
main.js:10 [ffmpeg-stderr] [aac @ 0x12854c0] Qavg: 122.576
main.js:10 [ffmpeg-stdout] FFMPEG_END
main.js:10 [info] FS.readFile E#######m-1#########8.mkv
localhost/:1 Access to fetch at 'https://s#######a.kinesisvideo.eu-central-1.amazonaws.com/putMedia' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
main.js:301 POST https://s#######a.kinesisvideo.eu-central-1.amazonaws.com/putMedia net::ERR_FAILED`

I thought that maybe this example need to be deployed on AWS to run it on the same domain as KVS(https://s#####a.kinesisvideo.eu-central-1.amazonaws.com) but the description clearly states about running it on localhost: - "http-server -a localhost -p 8080 -o"

Could anyone help with this issue?

Thank you

**Comment by MuhammedBeyhanUS** [comment]
hi @darek-gc ,
it says "No 'Access-Control-Allow-Origin' header is present on the requested resource."
As much as i see the problem stems from header. You have to add something like Access-Control-Allow-Origin  * (star for all) in your request header.  Other than i can not see any problem.

If running it on local is for only development/test purposes, you can also use some google extension that provides you proxy that lets you get over the CORS problem.  The best i found is this [https://chrome.google.com/webstore/detail/allow-cors-access-control/lhobafahddgcelffkeicbaginigeejlf](url) . 

Have a healthy life.

**Comment by darek-gc** [comment]
Thank you for your answer @MuhammedBeyhanUS 

With chrome extension you suggested error looks a little different:
```
...
[ffmpeg-stderr] [aac @ 0x1545e80] Qavg: 123.419
main.js:10 [ffmpeg-stdout] FFMPEG_END
main.js:10 [info] FS.readFile ExampleStream-1638994500715.mkv
localhost/:1 Access to fetch at 'https://s-96607c7a.kinesisvideo.eu-central-1.amazonaws.com/putMedia' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.
main.js:301 POST https://s-96607c7a.kinesisvideo.eu-central-1.amazonaws.com/putMedia net::ERR_FAILED
```

there is a problem with the OPTIONS request:
![image](https://user-images.githubusercontent.com/86299738/145280462-9bbca7b5-a68d-41fd-aa27-6e2827250a6a.png)

![image](https://user-images.githubusercontent.com/86299738/145280743-18ab38a7-0981-44e1-b3ca-fab52bcca98b.png)

Thanks

**Comment by niyatim23** [comment]
Another solution to avoid this would be to use an EC2 instance as your server. 

Closing this issue since the discussion is stale. Feel free to reopen this / open a new issue
